In [97]:
!pip3 install pandas

# Manipulating Data with Pandas! <br><img src = 'https://files.worldwildlife.org/wwfcmsprod/images/Panda_in_Tree/hero_small/99i33zyc0l_Large_WW170579.jpg'>
<h3> <a href= 'https://www.example.com'>Please refer to this dataset</a> as we will be using this dataset for the exercises below </h3>

<h3> Exercise 1a. </h3>
<p> When you open up the dataset, it looks quite dirty, with many rows having the order number replaced by "Imagine having no order number" etc. </p> 

1. Your first task would be to clear all rows with empty values in ```'PRODUCTCODE', 'ADDRESSLINE1'```,
2. avoid clearing the rows with empty values in ```'STATE'```,
3. and replace all the empty values under ```'ORDERNUMBER'``` with  ```10900 + index``` to give the row a unique ```'ORDERNUMBER'```,
4. replace all the empty values under ```'STATUS'``` with ```'To check'```
5. Lastly, save into a new file called ```'sales_data_cleaned.csv'```</p>

In [98]:
# Your code here

import pandas as pd

sales_data_dirty = pd.read_csv('./sales_data_dirty.csv')


sales_data_dirty.dropna(subset = ['PRODUCTCODE', 'ADDRESSLINE1'], axis= 0, inplace=True)
sales_data_dirty['STATUS'].fillna('To check', inplace=True)

# sales_data_dirty['ORDERNUMBER'].fillna(10000, inplace= True)
for index, row in sales_data_dirty.iterrows():
    if pd.isna(row['ORDERNUMBER']):
        sales_data_dirty.loc[index, 'ORDERNUMBER'] = 10900 + index

sales_data_dirty.head()



sales_data_dirty.drop('Unnamed: 0', axis= 1).to_csv('./sales_data_cleaned.csv')

<h3> Congratulations on finishing your first task! </h3>
<h4> Your csv file should look like this: </h4>

<img src='https://github.com/python-noob-dotcom/Computing-notes/raw/8c1416fd57990ed35931c7db361b3036f75810df/Practices/bbcs_testing/dataset_exercise_1.png'>

<h4> as the values are clean now </h4>

While cleaning the data, you noticed that there are some rows with a missing price each value

### Exercise 1b:
1. Using the ```.agg()``` method, retrieve the mean ```'PRICEEACH'```,
2. and replace the empty values in the row with the mean

In [99]:
# Your code here

sales_data = pd.read_csv('./sales_data_cleaned.csv', encoding= 'utf-8')

sales_data['PRICEEACH'].fillna(sales_data['PRICEEACH'].mean(), inplace=True)

sales_data.to_csv('./sales_data_cleaned.csv')


<h3> Exercise 2.</h3>
<p> Your boss then asks you to group the rows with the same STATUS together</p>
Group the rows with the status

```'Shipped', 'Disputed', etc together.```

In [100]:
# Your code here

import pandas as pd

sales_data = pd.read_csv('./sales_data_cleaned.csv', encoding= 'utf-8')
sales_data['STATUS'] = sales_data['STATUS'].astype(str)
grouped = sales_data.groupby('STATUS')
sales_data_grouped = sales_data
i = 0
for name, group in grouped:
    for index, row in group.iterrows():
        sales_data_grouped.iloc[i] = row
        i += 1

# sales_data_grouped_by_status = sales_data.groupby(['STATUS'], as_index= False).apply(lambda x: x)
sales_data_grouped.drop(['Unnamed: 0', 'Unnamed: 0.1'], axis= 1).to_csv('./sales_data_grouped.csv')

### Congratulations on completing your second exercise! Your dataset should look like this:

<image src = ''>

<img src = 'https://github.com/python-noob-dotcom/Computing-notes/blob/main/Practices/bbcs_testing/dataset_exercise_2.png?raw=true' alt= 'Answer to exercise 2'>

### Exercise 3

Create a new column `'Total revenue'` containing the total amount of revenue (quantity * price) collected by the company

In [101]:
# Your code here
ttl_revenue = 0

for index, row in sales_data.iterrows():
    if row['PRICEEACH'] == float('NaN') or row['QUANTITYORDERED'] == float('NaN'):
        continue

    ttl_revenue += int(row['QUANTITYORDERED']) * int(row['PRICEEACH'])

sales_data_with_ttl_revenue = sales_data
sales_data_with_ttl_revenue['Total revenue'] = ttl_revenue
print(ttl_revenue)

2201725


You should have gotten 2201725 as the total revenue

### Exercise 4

Find the top 5 top selling products based on the total sales (quantity sold)

In [102]:
# Your code here

product_sales = sales_data.groupby('PRODUCTLINE')['QUANTITYORDERED'].sum().reset_index()

top_selling = product_sales.sort_values(by = 'QUANTITYORDERED', ascending = False)

top_selling.head()

PRODUCTLINE  QUANTITYORDERED
0      Classic Cars            11185
4      Vintage Cars             4047
3  Trucks and Buses             4010
1       Motorcycles             3744
2            Planes             1295

#### You should have gotten the top 5 as 
1. Classic Cars
2. Vintage Cars
3. Trucks and Buses
4. Motorcycles
5. Planes

<img src = 'https://github.com/python-noob-dotcom/Computing-notes/blob/main/Practices/bbcs_testing/dataset_exercise_4.png?raw=true' alt = 'Exercise 4 answers'>